### PR TITLE
2D Sprite Support

### PR DESCRIPTION
**Pull request by vkalpias**
This PR adds the following for 2D Sprite support:

* Sprite Atlas Asset: Essentially a texture with extra data to define frames in the atlas
* Sprite Asset: An asset that references the Sprite Atlas asset and an array of keys that reference frames from that atlas
* `pc.SpriteAtlas`: The resource of the Sprite Atlas Asset
* `pc.Sprite`: The resource of the Sprite Asset
* `pc.SpriteComponent`: The component that renders 'simple' or 'animated' sprites
* `pc.SpriteAnimationClip`: Handles loading and playing of a sprite asset - can also be used to render a single frame and it used internally by simple SpriteComponents too to render a single frame.

### CODE REVIEW

File: src/scene/sprite-atlas.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    'use strict';
+
+    /**
+    * @name pc.SpriteAtlas
+    * @class Represents the resource of a sprite atlas asset.
+    * @property {pc.Texture} texture The texture atlas.
+    * @property {Object} frames Contains frames which define portions of the texture atlas.
+    * @example
+    * var atlas = new pc.SpriteAtlas();
+    * atlas.frames = {
```
Comment (willeastcott) [inline src/scene/sprite-atlas.js]: This may be a dumb question, but why is this an object and not an array?

Comment (vkalpias) [reply src/scene/sprite-atlas.js]: Frames have a unique key and if you add or remove frames in between of other frames in the atlas the keys of the existing frames shouldn't change. So this is an easier way to change the atlas without having to change all the sprite assets that reference these frames.

Comment (guycalledfrank) [reply src/scene/sprite-atlas.js]: I recently noticed that for...in loops can be quite slow, so maybe some other solution (emulate linked list with 2 arrays?) will be better

File: src/framework/components/sprite/sprite-animation-clip.js [hunk 2]
```diff
@@ -0,0 +1,494 @@
...
+        this._time = 0;
+
+        pc.events.attach(this);
+    };
+
+    SpriteAnimationClip.prototype = {
+        _createMeshes: function () {
+            var i;
+
+            // destroy old meshe
```
Comment (H1Gdev) [inline src/framework/components/sprite/sprite-animation-clip.js]: // destroy old meshes

File: src/scene/sprite-atlas.js (hunk 1 above)
Comment (vkalpias) [reply src/scene/sprite-atlas.js]: We don't loop through this object. The sprites assets reference keys of this object directly.

File: src/resources/sprite-atlas.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    var JSON_FILTER_MODE = {
+        "nearest":             pc.FILTER_NEAREST,
+        "linear":              pc.FILTER_LINEAR,
+        "nearest_mip_nearest": pc.FILTER_NEAREST_MIPMAP_NEAREST,
+        "linear_mip_nearest":  pc.FILTER_LINEAR_MIPMAP_NEAREST,
+        "nearest_mip_linear":  pc.FILTER_NEAREST_MIPMAP_LINEAR,
+        "linear_mip_linear":   pc.FILTER_LINEAR_MIPMAP_LINEAR
+    };
+
+    var SpriteAtlasHandler = function (loader) {
```
Comment (daredevildave) [inline src/resources/sprite-atlas.js]: This should be `TextureAtlas` not Sprite Atlas.

File: src/scene/sprite-atlas.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
+pc.extend(pc, function () {
+    'use strict';
+
+    /**
+    * @name pc.SpriteAtlas
```
Comment (daredevildave) [inline src/scene/sprite-atlas.js]: This should be TextureAtlas not SpriteAtlas

File: src/scene/sprite-atlas.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    *   '1': {
+    *       rect: new pc.Vec4(0.5, 0.5, 0.5, 0.5),
+    *       pivot: new pc.Vec2(0.5, 0.5)
+    *   },
+    *   ...
+    * };
+    */
+    var SpriteAtlas = function () {
+        this.texture = null;
+        this.frames = null;
```
Comment (daredevildave) [inline src/scene/sprite-atlas.js]: Should this be an empty array?

File: src/scene/sprite.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    * @name pc.Sprite
+    * @class Represents the resource of a sprite asset.
+    * @property {Number} pixelsPerUnit The number of pixels that map to one PlayCanvas unit.
+    * @property {pc.SpriteAtlas} atlas The sprite atlas.
+    * @property {String[]} frameKeys The keys of the frames in the sprite atlas that this sprite is using.
+    */
+    var Sprite = function () {
+        this.pixelsPerUnit = 1;
+        this.atlas = null;
+        this.frameKeys = null;
```
Comment (daredevildave) [inline src/scene/sprite.js]: Should this be an empty array?

File: src/scene/sprite-atlas.js (hunk 4 above)
Comment (willeastcott) [reply src/scene/sprite-atlas.js]: Good point, @daredevildave. I do agree. However, a texture atlas is quite a generic thing. Is the SpriteAtlas class currently geared just towards sprites? Looks like it is. Does this object belong in the src/graphics folder? Or src/scene? Comments @guycalledfrank?

Comment (daredevildave) [reply src/scene/sprite-atlas.js]: It's pretty general purpose, it defines a Texture that has uv rects defined.

Comment (vkalpias) [reply src/scene/sprite-atlas.js]: So it is 'kinda' general purpose. A frame has uv's but it also have pivots which are unrelated outside the context of sprites. That's why I named it SpriteAtlas. We could use it for regular materials too I guess but then we'd also store the pivot which is probably not needed.

File: src/scene/sprite-atlas.js (hunk 5 above)
Comment (vkalpias) [reply src/scene/sprite-atlas.js]: It's a dictionary. We could initialize it as a dictionary although that's gonna be garbage collected once we load the real frames unless we always modify the same object...

File: src/scene/sprite.js (hunk 6 above)
Comment (vkalpias) [reply src/scene/sprite.js]: Same as with the SpriteAtlas#frames

**Pull request merged**